Networking/Networking/ProductImageInBackground/ProductImageStatusStorage.swift

```diff
@@ -0,0 +1,183 @@
+import Foundation
+import Combine
+
+/// Save product image upload statuses in User Defaults.
+/// This class is declared in the Networking layer because it will also be accessed by the background URLSession operations.
+/// This class avoid KVO, and uses Combine to handle notifications, and efficiently checks for actual data changes.
+/// We're not using KVO, because `ProductImageStatus` use a swift type (enum), not compatible with Obj-C.
+///
+public final class ProductImageStatusStorage {
+    private let key: String
+    private let userDefaults: UserDefaults
+
+    // Private internal instance for static methods
+    private static let internalInstance = ProductImageStatusStorage()
+
+    // Publisher for observing changes
+    private let statusesSubject: CurrentValueSubject<[ProductImageStatus], Never>
+    private var cancellables = Set<AnyCancellable>()
+
+    private let encoder: JSONEncoder
+    private let decoder: JSONDecoder
+
+    public var statusesPublisher: AnyPublisher<[ProductImageStatus], Never> {
+        statusesSubject
+            .receive(on: DispatchQueue.main)
+            .eraseToAnyPublisher()
+    }
+
+    public var errorsPublisher: AnyPublisher<(siteID: Int64,
+                                              productOrVariationID: ProductOrVariationID?,
+                                              assetType: ProductImageAssetType?,
+                                              error: Error), Never> {
+        statusesSubject
+            .compactMap { statuses in
+                statuses.first { $0.isUploadFailure }
+            }
```

Comment:
❓ I wonder what this publisher is going to be used for? Why do we only need to get one single error instead of all the errors in the upload list?

Reply:
During the implementation, I considered that from a user's perspective, higher-level error handling might be clearer if we only processed the first error in the list. However, now that you've asked, I realize we use this approach to handle all errors for each image. Therefore, it makes sense to return all the error statuses, as they will be addressed at the UI level. 👍

Reply:
I implemented the changes in this commit: 016d77740fd7e27bc6d0604e41c898ffbdf1073d